spec/unit/settings_spec.rb

```diff
@@ -287,6 +287,58 @@ def metadata(setting)
       expect(@settings.set_by_cli?(:myval)).to be_falsey
     end
 
+    it "should find no configured settings by default" do
+      expect(@settings.set_by_config?(:myval)).to be_falsey
+    end
+
+    it "should identify configured settings in memory" do
+      @settings.instance_variable_get(:@value_sets)[:memory].expects(:lookup).with(:myval).returns('foo')
+      expect(@settings.set_by_config?(:myval)).to be_truthy
+    end
+
+    it "should identify configured settings from CLI" do
+      @settings.instance_variable_get(:@value_sets)[:cli].expects(:lookup).with(:myval).returns('foo')
+      expect(@settings.set_by_config?(:myval)).to be_truthy
+    end
+
+    it "should not identify configured settings from environment" do
```

Comment:
This one's interesting, but made sense for the methods use in setting agent_specified_environment. We wouldn't expect the environment.conf to specify an environment.

In general, it would make sense to extend set_by_config? to take an environment argument as an option.